lib/history/sqlite.go

```diff
@@ -0,0 +1,224 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package history
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"strings"
+	"time"
+
+	pb "github.com/gravitational/satellite/agent/proto/agentpb"
+
+	"github.com/gravitational/trace"
+	_ "github.com/mattn/go-sqlite3" // initialize sqlite3
+)
+
+// SQLiteTimeline represents a timeline of cluster status events. The timeline
+// can hold a specified amount of events and uses a FIFO eviction policy.
+// Timeline events are stored in a local sqlite database.
+//
+// Implements Timeline
+type SQLiteTimeline struct {
+	// size specifies the max number of events stored in the timeline.
+	size int
+	// database points to underlying sqlite database.
+	database *sql.DB
+	// lastStatus holds the last recorded cluster status.
+	lastStatus *Cluster
+}
+
+// NewSQLiteTimeline initializes and returns a new SQLiteTimeline with the
+// specified size. Initial cluster status is `Unknown`.
+func NewSQLiteTimeline(database *sql.DB, size int) (Timeline, error) {
+	if err := initEventTable(database); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &SQLiteTimeline{
+		size:       size,
+		database:   database,
+		lastStatus: &Cluster{Status: pb.SystemStatus_Unknown.String()},
+	}, nil
+}
+
+// initEventTable initializes event table schema.
+func initEventTable(db *sql.DB) error {
+	if _, err := db.Exec(createTableEvent); err != nil {
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+// RecordStatus records differences of the previous status to the provided
+// status into the Timeline.
+func (t *SQLiteTimeline) RecordStatus(ctx context.Context, status *pb.SystemStatus) error {
+	cluster := parseSystemStatus(status)
+	events := t.lastStatus.diffCluster(cluster)
+	if len(events) == 0 {
+		return nil
+	}
+
+	tx, err := t.database.BeginTx(ctx, nil)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	if err := t.insertEvents(events); err != nil {
+		tx.Rollback()
+		return trace.Wrap(err, "failed to insert events.")
+	}
+
+	if err := t.evictEvents(); err != nil {
+		tx.Rollback()
+		return trace.Wrap(err, "failed to evict old events.")
+	}
+
+	if err := tx.Commit(); err != nil {
+		return trace.Wrap(err)
+	}
+
+	t.lastStatus = cluster
+	return nil
+}
+
+// GetEvents returns the current timeline.
+func (t *SQLiteTimeline) GetEvents() (events []*Event, err error) {
+	rows, err := t.database.Query(selectAllFromEvent)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	var (
+		id        int
+		timestamp time.Time
+		eventType EventType
+		node      string
+		probe     string
+		old       string
+		new       string
+	)
+
+	for rows.Next() {
+		if err := rows.Scan(&id, &timestamp, &eventType, &node, &probe, &old, &new); err != nil {
```

Comment:
Here's a helpful extension for working on a struct level: https://godoc.org/github.com/jmoiron/sqlx#Rows.StructScan if it ever grows in complexity.